File: YABA-New/YABA/app/src/main/java/dev/subfly/yaba/core/model/annotation/ReadableSelectionDraft.kt
package dev.subfly.yaba.core.model.annotation

import androidx.compose.runtime.Stable
import kotlinx.serialization.Serializable

/**
 * Selection payload from a WebView for annotation creation.
 * - Rich-text reader: [pdfAnchor] is null; quote-only (positions live in TipTap `yabaAnnotation` marks after apply).
 * - PDF viewer: [pdfAnchor] holds section offsets stored in DB [AnnotationEntity.extrasJson].
 * - EPUB viewer: [epubAnchor] holds CFI range in [AnnotationEntity.extrasJson].
 */
@Serializable
@Stable
data class ReadableSelectionDraft(
    val sourceContext: AnnotationSourceContext,
    val quote: AnnotationQuoteSnapshot,
    val pdfAnchor: PdfAnnotationExtras? = null,
    val epubAnchor: EpubAnnotationExtras? = null,
) {
    val bookmarkId: String get() = sourceContext.bookmarkId
    val readableVersionId: String get() = sourceContext.contentId
}

File: Darwin/YABACore/Model/Annotation/ReadableSelectionDraft.swift
//
//  ReadableSelectionDraft.swift
//  YABACore
//
//  Contract-only draft passed from web/PDF/EPUB hosts for annotation creation.
//

import Foundation

public struct ReadableSelectionDraft: Sendable, Equatable {
    public var bookmarkId: String
    public var quoteText: String?
    /// Optional JSON for anchors (PDF offsets, EPUB CFI, etc.).
    public var extrasJson: String?
    /// Source annotation surface (`READABLE`, `PDF`, `EPUB`).
    public var annotationType: AnnotationType

    public init(
        bookmarkId: String,
        quoteText: String? = nil,
        extrasJson: String? = nil,
        annotationType: AnnotationType = .readable
    ) {
        self.bookmarkId = bookmarkId
        self.quoteText = quoteText
        self.extrasJson = extrasJson
        self.annotationType = annotationType
    }

    public var sourceContext: AnnotationSourceContext {
        switch annotationType {
        case .readable:
            return .readable(bookmarkId: bookmarkId)
        case .pdf:
            return .pdf(bookmarkId: bookmarkId)
        case .epub:
            return .epub(bookmarkId: bookmarkId)
        }
    }

    public var quote: AnnotationQuoteSnapshot {
        AnnotationQuoteSnapshot.fromSelectedText(quoteText ?? "")
    }
}
